adareducer/engine.py

```diff
@@ -194,14 +219,31 @@ def next_file_to_process(self):
     def run(self):
         """Run self: reduce the project as much as possible"""
 
+        all_files = set([os.path.abspath(f) for f in os.listdir(self.project_dir)])
+
         # Before running any modification, run the predicate,
         # as a sanity check.
         if not self.run_predicate(True):
             log("The predicate returned nonzero")
             return
 
+        predicate_artifacts = (
+            set([os.path.abspath(f) for f in os.listdir(self.project_dir)]) - all_files
+        )
+
         # We've passed the sanity check, time to reduce!
 
+        # Attempt to remove all files in the directory before doing any
+        # reduction: this might save time by deleting files we would have tried
+        # to reduce.
+        if DELETE_FILES_BEFORE:
+            to_delete = [
+                f
+                for f in (all_files - predicate_artifacts - self.nonremovable_files)
```

Comment:
here we would need to protect from deletion any ".orig" file and any ".crash" file that could be generated by previous runs of adareducer - this is in the case that adareducer crashes and the user wants to salvage the run and fix manually.